File: cplusplus/src/sign.cpp
#include "sign.h"

#include "private.h"
#include "sizes.h"

namespace Autograph {

SignResult createErrorResult() {
  Bytes signature(SIGNATURE_SIZE, 0);
  return {false, signature};
}

SignFunction createSafeSign(const SignFunction sign) {
  SignFunction safeSign = [sign](const Bytes subject) {
    try {
      auto signResult = sign(subject);
      if (signResult.signature.size() != SIGNATURE_SIZE) {
        return createErrorResult();
      }
      return signResult;
    } catch (...) {
      return createErrorResult();
    }
  };
  return safeSign;
}

SignFunction createSign(const Bytes identityPrivateKey) {
  SignFunction sign = [identityPrivateKey](Bytes subject) {
    Bytes signature(SIGNATURE_SIZE);
    bool success =
        autograph_sign_subject(signature.data(), identityPrivateKey.data(),
                               subject.data(), subject.size()) == 0;
    SignResult result = {success, signature};
    return result;
  };
  return sign;
}

}  // namespace Autograph

File: typescript/src/sign.ts
import { autograph_sign_subject } from './clib'
import { SignFunction, SignResult } from '../types'
import { SIGNATURE_SIZE, createSignatureBytes } from './utils'

const createErrorSignResult = (): SignResult => ({
  success: false,
  signature: new Uint8Array(SIGNATURE_SIZE)
})

const ensureSignResult = (result: SignResult): SignResult => {
  if (result.signature.byteLength !== SIGNATURE_SIZE) {
    return createErrorSignResult()
  }
  return result
}

export const createSafeSign =
  (sign: SignFunction): SignFunction =>
  async (subject: Uint8Array) => {
    try {
      const result = await sign(subject)
      return ensureSignResult(result)
    } catch (error) {
      return createErrorSignResult()
    }
  }

export const createSign =
  (identityPrivateKey: Uint8Array): SignFunction =>
  (subject: Uint8Array) => {
    const signature = createSignatureBytes()
    const success = autograph_sign_subject(
      signature,
      identityPrivateKey,
      subject,
      subject.byteLength
    )
    return { success, signature }
  }
